model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -587,6 +587,12 @@ public void preRemove(RealmModel realm, RoleModel role) {
         // GroupProvider method implementation ends here
     }
 
+    @Override
+    public void preRemoveClientProvider(RealmModel realm, RoleModel role) {
+        String clientScopeMapping = JpaUtils.getTableNameForNativeQuery("SCOPE_MAPPING", em);
+        em.createNativeQuery("delete from " + clientScopeMapping + " where ROLE_ID = :role").setParameter("role", role.getId()).executeUpdate();
```

Comment:
Could this be a JPA named query?

Reply:
I thikn it's not possible using a delete. The reason is that this table is a [collection table](https://github.com/keycloak/keycloak/blob/12.0.4/model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java#L161-L164). It's the same that happens [with CHILD_ROLE a few lines above](https://github.com/keycloak/keycloak/blob/12.0.4/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java#L350). I can use a named query to obtain the list of clients that have the scope mapping (that is done using an inner join) and perform the deletion (`deleteScopeMapping` method) one by one. I decided to do it this way because it is much better (one query instead of N+1) and because it was already done with the CHILD_ROLE. But if you prefer this I can change to that approach.